kong/plugins/http-log/handler.lua

```diff
@@ -72,36 +48,47 @@ local function log(premature, conf, body, name)
   end
   name = "[" .. name .. "] "
 
-  local ok, err
+  local headers
+  local req_url
   local parsed_url = parse_url(conf.http_endpoint)
-  local host = parsed_url.host
-  local port = tonumber(parsed_url.port)
 
-  local sock = ngx.socket.tcp()
-  sock:settimeout(conf.timeout)
-
-  ok, err = sock:connect(host, port)
-  if not ok then
-    ngx.log(ngx.ERR, name .. "failed to connect to " .. host .. ":" .. tostring(port) .. ": ", err)
-    return
+  if parsed_url.userinfo then
+    -- for lua-resty-http client Authorization
+    headers = {
+      ["Content-Type"] = conf.content_type,
+      ["Authorization"] = string_format(
+        "Basic %s", ngx.encode_base64(parsed_url.userinfo))
+    }
+    req_url = parsed_url.scheme .. "://" .. parsed_url.host .. ":" .. parsed_url.port
+      .. parsed_url.path .. (parsed_url.params or "")
+  else
+    headers = {
+      ["Content-Type"] = conf.content_type
+    }
+    req_url = conf.http_endpoint
   end
 
-  if parsed_url.scheme == HTTPS then
-    local _, err = sock:sslhandshake(true, host, false)
-    if err then
-      ngx.log(ngx.ERR, name .. "failed to do SSL handshake with " .. host .. ":" .. tostring(port) .. ": ", err)
-    end
+  local httpc = http.new()
+  httpc:set_timeout(conf.timeout)
+  local res, err = httpc:request_uri(req_url, {
+      ssl_verify = false,
+      method = conf.method:upper(),
+      body = body,
+      headers = headers
+  })
+
+  if not res then
+    ngx.log(ngx.ERR, name .. "request failed: " .. req_url, err)
+    return
   end
 
-  ok, err = sock:send(generate_post_payload(conf.method, conf.content_type, parsed_url, body))
-  if not ok then
-    ngx.log(ngx.ERR, name .. "failed to send data to " .. host .. ":" .. tostring(port) .. ": ", err)
+  if res.status >= 400 then
+    ngx.log(ngx.ERR, name .. "request failed: " .. req_url .. "status: " .. tostring(res.status), err)
   end
 
-  ok, err = sock:setkeepalive(conf.keepalive)
+  local ok, err = httpc:set_keepalive(conf.keepalive)
```

Comment:
The [request_uri()](https://github.com/pintsized/lua-resty-http/blob/master/lib/resty/http.lua#L792) interface does the `set_keepalive()` automatically for us, so technically this would not be necessary.

However, using the `request_uri()` interface also renders the `conf.keepalive` parameter useless. We should probably use the lower-level `request()` interface instead. This also mean we will need to handle the SSL handshake ourselves, something that the `request_uri()` interface was doing for us.

Reply:
I PR for http-lua-resty add an option for keepalive, I think that way simple and grace.

Reply:
That works as well, unfortunately this PR will have to stay stale until/if it is merged upstream and Kong updates its dependency.

Reply:
for reference, this is the PR: https://github.com/pintsized/lua-resty-http/pull/145

Reply:
would it be too ugly to peek inside `httpc` and do `httpc.sock:set_keepalive` in the mean time?

Reply:
Yeah, let's avoid going down that road... If this has been undetected for so long it can stay like this a little bit longer and avoid us setting dangerous precedents.